concent_api/core/message_handlers.py

```diff
@@ -228,6 +230,18 @@ def handle_send_reject_report_computed_task(client_message):
             "RejectReportComputedTask cannot contain CannotComputeTask and TaskFailure at the same time.",
             error_code=ErrorCode.MESSAGE_INVALID,
         )
+    try:
+        validate_if_given_object_is_expected_enum_instance(client_message.reason, message.RejectReportComputedTask.REASON)
+    except NoneAsValueNotAllowedError:
+        raise GolemMessageValidationError(
```

Comment:
I think this way of defining validation functions is not the best way to do it. You have created a validator to move some code out of the handler but then you had to add a big `try ... except` block. Overall it's more code than if you did it the simplest way and it is not even reusable. To validate it in another handler you'd have to copy the whole `try ... except` block.

I'd replace it with something simpler. A function called `validate_reject_report_computed_task(message)` that simply does
``` python
if isinstance(message.reason, message.RejectReportComputedTask.REASON):
    raise GolemMessageValidationError(...)
```

That's simple, does have enough context to construct an error message (it knows where the enum values comes from) and is easily extensible (we can add more checks to this function later; I'd actually put the `if` that's a few lines above there too). That's basically how our validations looked like at the beginning of the project. I think it was good enough for a simple validation system and the only problem with it was that it used `Http400` rather than an exception that represents a validation error.

If you really want to make it generic and reusable, I'd create something similar to Django's validator system. Something where you can define validators for each field and then have some function run the validators, catch exceptions and prepare error messages based on the name of the field and the message from the exception. Here's how it could work in our case:
1. Create a class called `MessageValidator`.
    - Has a validator map: field name -> list of validation functions, e.g.
    - Has a class method called `validate(message)`.
        - This method goes over all fields listed in the map and for each runs all the corresponding validators.
        - If a validator function raises `GolemMessageValidationError` it takes the error message, combines it with the name of the field and name of the message class and then raises the error again with the combined message.
2. Derive a specific class for each message type. For example `RejectReportComputedTaskValidator`.
    - Defines validation functions for all fields that need them
    - Each function is added to the validator map using `@register_validator(field_name)` decorator.
        - A validation function takes field value as argument and either returns or raises `GolemMessageValidationError`.
3. Create a decorator called `@register_validator(field_name)`.
    - Adds the class method to the validator map under the specified field name.
    - If the name is not specified, the validator is for the whole message and the function will get the message instead of just a value of a single field.
4. It should also be possible to register another validator class for a specific field (when it's a nested message).

But I'd suggest doing something like this in a separate issue. To finish this pull request quickly I'd go with the simple one I suggested above.

@rwrzesien What do you think about a system of validators like this?

Reply:
@cameel Shouldn't it be ?
```
if not isinstance(message.reason, message.RejectReportComputedTask.REASON):
    raise GolemMessageValidationError(...)
```

My idea was to propose of using Django Forms (with probably several tweaks) for validation as soon as we will have each Golem message stored in separate Model. We can then gather all errors from message, not just first. But it doesn't look like it is going to be soon. However, I think your idea describes a similar mechanism. 

How I see it is:
- Define atomic generic validators.
- Create some base validation class handling validation flow.
- Define a specific class for each message type, as you have proposed. This function will have a list of fields defined as assigned a list of validators to them.

I would still check if any part of Django Forms can be used for doing that. This way if someday we will have separate Models for Golem messages we can easily switch/add Django Forms, which is more natural for Django Model validation.

Reply:
> Shouldn't it be ?
> ```
> if not isinstance(message.reason, message.RejectReportComputedTask.REASON):
>     raise GolemMessageValidationError(...)
> ```

Yes. My mistake.

> However, I think your idea describes a similar mechanism.
> 
> How I see it is:

Yes. It was just an example of such a mechanism. I did not want to suggest Django Forms because they're fine overall but I think we're going to have lots of small problems with details like differences between `""` and `None`. Django Forms is pretty lax about that because it assumes that the data comes from a HTML form anyway. I'd look for a different library if I were to implement this myself.

But anyway, my point was that the simple mechanism we're using from the beginning where we have:
- a validation function for each message type which checks a series of conditions and is responsible for creating error messages and raising errors
- helper functions that check more complex conditions and only return `True`/`False`

is still good enough. It's better than creating reusable, nested validators ad-hoc because such validators don't have enough context without a bigger mechanism that puts it all together. So I'd stick with it until we get something better.

And as for something better - I'd leave it up to you. Use whatever you like, even Django Forms if you think you can make something good out of it. My only requirement would be that I'd like to avoid it growing to the scale of a big task that is constantly being pushed back like it happened with Django checks. I'd like you to design something on a smaller scale that can be implemented immediately and then over time extend it to make it better.

Reply:
>  I did not want to suggest Django Forms because they're fine overall but I think we're going to have lots of small problems with details like differences between "" and None.

I agree, but those are the parts that I think we can override and prepare for golem-messages. I am not saying that it must be Django Forms, I just think that they might have some parts of the wheel that we can use.

Lets discuss it with @dybi too when he gets back.